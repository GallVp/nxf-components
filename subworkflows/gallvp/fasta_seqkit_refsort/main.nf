include { SEQKIT_SORT                       } from '../../../modules/gallvp/seqkit/sort/main'
include { MINIMAP2_ALIGN                    } from '../../../modules/gallvp/minimap2/align/main'
include { JUICEBOXSCRIPTS_MAKEAGPFROMFASTA  } from '../../../modules/gallvp/juiceboxscripts/makeagpfromfasta/main'
include { HAPHIC_REFSORT                    } from '../../../modules/gallvp/haphic/refsort/main'
include { CUSTOM_INTERLEAVEFASTA            } from '../../../modules/gallvp/custom/interleavefasta/main'

workflow FASTA_SEQKIT_REFSORT {

    take:
    ch_fasta                                // channel: [ val(meta), fasta ]
    val_fasta_combinations                  // val: []|null|"x x:y a:b ..."
    val_alphanumeric_sort                   // val: true|false
    val_refsort                             // val: true|false

    main:

    ch_versions                             = Channel.empty()

    // MODULE: SEQKIT_SORT; ext.args = '--ignore-case --natural-order'
    SEQKIT_SORT ( val_alphanumeric_sort ? ch_fasta : Channel.empty() )

    ch_sorted_fasta                         = val_alphanumeric_sort
                                            ? SEQKIT_SORT.out.fastx
                                            : ch_fasta
    ch_versions                             = ch_versions.mix(SEQKIT_SORT.out.versions.first())

    // Channels: ch_grouped_fastas, ch_single_fasta, ch_paired_fastas, ch_minimap_inputs
    ch_combinations                         = ( val_fasta_combinations == null || val_fasta_combinations == [] )
                                            ? ch_fasta
                                            | map { meta, _fasta -> meta.id }
                                            : Channel.of(
                                                val_fasta_combinations.tokenize( ' ' )
                                            )
                                            | flatten

    ch_grouped_fastas                       = ch_sorted_fasta
                                            | combine ( ch_combinations )
                                            | filter { meta, _fasta, comb -> meta.id in comb.tokenize( ':' ) }
                                            | map { meta, fasta, comb ->
                                                [ comb, meta + [ comb: comb ], fasta ] // meta2: meta + [comb:]
                                            }
                                            | groupTuple
                                            | branch { _comb, _metas2, fastas ->
                                                single: fastas.size() == 1
                                                multi: fastas.size() > 1
                                            }

    ch_single_fasta                         = ch_grouped_fastas.single
                                            | map { _comb, metas2, fastas ->
                                                [ metas2.first(), fastas.first() ]
                                            }

    ch_paired_fastas                        = ch_grouped_fastas.multi
                                            | flatMap { comb, metas2, fastas ->

                                                def comb_ids = comb.tokenize(':') // ordered
                                                def meta_ids = metas2.collect { it.id } // unordered

                                                def ref_id = comb_ids.last()
                                                def query_ids = comb_ids[0..-2]

                                                def ref_fasta = fastas[ meta_ids.indexOf( ref_id ) ]

                                                query_ids.collect { it ->
                                                    [
                                                        id: "${it}_${ref_id}",
                                                        comb: comb,
                                                        pair: "${it}:${ref_id}",
                                                        fastas: [ fastas[ meta_ids.indexOf( it ) ], ref_fasta ]
                                                    ]
                                                }
                                            }
                                            | map { box ->
                                                [
                                                    [
                                                        id: box.id,
                                                        comb: box.comb,
                                                        pair: box.pair
                                                    ],
                                                    box.fastas
                                                ]
                                            }


    // MODULE: MINIMAP2_ALIGN; ext.args = -x asm5 --secondary=no
    ch_minimap_inputs                       = val_refsort
                                            ? ch_paired_fastas
                                            : Channel.empty()
    MINIMAP2_ALIGN (
        ch_minimap_inputs.map { meta2, fastas -> [ meta2, fastas.first() ] },
        ch_minimap_inputs.map { meta2, fastas -> [ meta2, fastas.last() ] },
        false,  // bam_format
        [],     // bam_format
        false,  // cigar_paf_format
        false   // cigar_bam
    )

    ch_minimap_inputs_paf                   = ch_minimap_inputs
                                            | join(MINIMAP2_ALIGN.out.paf)
    ch_versions                             = ch_versions.mix(MINIMAP2_ALIGN.out.versions.first())

    // JUICEBOXSCRIPTS_MAKEAGPFROMFASTA
    JUICEBOXSCRIPTS_MAKEAGPFROMFASTA (
        ch_minimap_inputs.map { meta2, fastas -> [ meta2, fastas.first() ] }
    )

    ch_minimap_inputs_paf_agp               = ch_minimap_inputs_paf
                                            | join(
                                                JUICEBOXSCRIPTS_MAKEAGPFROMFASTA.out.agp
                                            )
    ch_versions                             = ch_versions.mix(JUICEBOXSCRIPTS_MAKEAGPFROMFASTA.out.versions.first())

    // MODULE: HAPHIC_REFSORT
    ch_refsort_inputs                       = ch_minimap_inputs_paf_agp
                                            | multiMap { meta2, fastas, paf, query_agp ->
                                                agp:    [ meta2, query_agp ]
                                                paf:    paf
                                                fasta:  fastas.first()
                                            }
    HAPHIC_REFSORT (
        ch_refsort_inputs.agp,
        ch_refsort_inputs.paf,
        ch_refsort_inputs.fasta,
        [] //ref_order
    )

    ch_minimap_inputs_refsort_fasta         = ch_minimap_inputs
                                            | join ( HAPHIC_REFSORT.out.fasta )
    ch_versions                             = ch_versions.mix(HAPHIC_REFSORT.out.versions.first())

    // Channel: ch_refsort_pairs
    ch_refsort_pairs                        = val_refsort
                                            ? ch_minimap_inputs_refsort_fasta
                                            | map { meta2, fastas, refsort_fasta ->
                                                [
                                                    meta2,
                                                    [ refsort_fasta, fastas.last() ]
                                                ]
                                            }
                                            : ch_paired_fastas

    // MODULE:CUSTOM_INTERLEAVEFASTA
    ch_interleave_input                     = ch_refsort_pairs
                                            | map { meta2, fastas ->
                                                [
                                                    meta2.comb,
                                                    meta2,
                                                    fastas
                                                ]
                                            }
                                            | groupTuple
                                            | multiMap { comb, metas2, fastas ->
                                                def unordered_fastas = fastas.collect { it.first() } + [ fastas.first().last() ]

                                                def unordered_ids = metas2.collect { meta -> meta.pair.tokenize(':').first() } + [ comb.tokenize(':').last() ]
                                                def ordered_ids = comb.tokenize(':')


                                                fastas: [
                                                        [ id: comb.tokenize(':').join('_'), comb: comb ],
                                                        ordered_ids.collect { id -> unordered_fastas[ unordered_ids.indexOf( id ) ] }
                                                    ]
                                                prefixes: ordered_ids.join( ' ' )
                                            }


    CUSTOM_INTERLEAVEFASTA (
        ch_interleave_input.fastas,
        ch_interleave_input.prefixes
    )

    ch_interleaved_fasta                    = CUSTOM_INTERLEAVEFASTA.out.fasta
    ch_versions                             = ch_versions.mix(CUSTOM_INTERLEAVEFASTA.out.versions.first())

    emit:
    fasta                                   = ch_interleaved_fasta
                                            | mix ( ch_single_fasta )   // channel: [ meta3, fasta ]; meta3 ~ [ id, comb ]
    versions                                = ch_versions               // channel: [ versions.yml ]
}
