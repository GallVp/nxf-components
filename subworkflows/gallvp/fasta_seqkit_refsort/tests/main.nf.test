nextflow_workflow {

    name "Test Subworkflow FASTA_SEQKIT_REFSORT"
    script "../main.nf"
    config "./nextflow.config"
    workflow "FASTA_SEQKIT_REFSORT"

    tag "subworkflows"
    tag "subworkflows_gallvp"
    tag "subworkflows/fasta_seqkit_refsort"

    tag "juiceboxscripts/makeagpfromfasta"
    tag "seqkit/sort"
    tag "haphic/refsort"
    tag "minimap2/align"
    tag "gunzip"
    tag "custom/interleavefasta"

    setup {
        run("GUNZIP") {
            script "../../../../modules/gallvp/gunzip/main.nf"
            process {
                """
                input[0] = Channel.of(
                    [
                        [ id: 'query' ],
                        file('https://github.com/Plant-Food-Research-Open/assemblyqc/raw/4f9fba392cf7212d61c6245fd72339e09164ce44/tests/hic/testdata/HYh1h2_Chr01_6000000_6015000.fasta.gz')
                    ],
                    [
                        [ id: 'ref' ],
                        file(params.modules_testdata_base_path + 'genomics/eukaryotes/actinidia_chinensis/genome/chr1/genome.fasta.gz', checkIfExists: true)
                    ],
                )
                """
            }
        }
    }

    test("HY - actinidia_chinensis - fasta") {

        when {
            workflow {
                """
                input[0] = GUNZIP.out.gunzip
                input[1] = null
                input[2] = true
                input[3] = true
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }


    test("HY - actinidia_chinensis - fasta - combinations") {

        when {
            workflow {
                """
                input[0] = GUNZIP.out.gunzip
                input[1] = "query query:ref ref"
                input[2] = true
                input[3] = true
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("HY - actinidia_chinensis - fasta - no - sort - combinations") {

        when {
            workflow {
                """
                input[0] = GUNZIP.out.gunzip
                input[1] = "query query:ref ref"
                input[2] = false
                input[3] = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}