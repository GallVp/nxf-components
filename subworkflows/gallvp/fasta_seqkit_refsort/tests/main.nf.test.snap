{
    "HY - actinidia_chinensis - fasta - no - sort - combinations": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "query_ref"
                        },
                        "query_ref.fasta:md5,7e52d0a17de86f5b11dd4d77c8332fa6"
                    ],
                    [
                        {
                            "id": "query"
                        },
                        "HYh1h2_Chr01_6000000_6015000.fasta:md5,698715c5989f43041a04f6fdca846974"
                    ],
                    [
                        {
                            "id": "ref"
                        },
                        "genome.fasta:md5,d07bc737a8ba2db4ea02872a73dace0b"
                    ]
                ],
                "1": [
                    "versions.yml:md5,5772594d0b3e1f41b2730f1dda23465b"
                ],
                "fasta": [
                    [
                        {
                            "id": "query_ref"
                        },
                        "query_ref.fasta:md5,7e52d0a17de86f5b11dd4d77c8332fa6"
                    ],
                    [
                        {
                            "id": "query"
                        },
                        "HYh1h2_Chr01_6000000_6015000.fasta:md5,698715c5989f43041a04f6fdca846974"
                    ],
                    [
                        {
                            "id": "ref"
                        },
                        "genome.fasta:md5,d07bc737a8ba2db4ea02872a73dace0b"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,5772594d0b3e1f41b2730f1dda23465b"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-07-02T10:18:34.182968"
    },
    "HY - actinidia_chinensis - fasta": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "query"
                        },
                        "query.fasta:md5,bcc14495671d77033b36be11fc20652c"
                    ],
                    [
                        {
                            "id": "ref"
                        },
                        "ref.fasta:md5,7315276c0e677b424e673562f48b9873"
                    ]
                ],
                "1": [
                    "versions.yml:md5,7cd4eb452407fcd946428f758c97442d"
                ],
                "fasta": [
                    [
                        {
                            "id": "query"
                        },
                        "query.fasta:md5,bcc14495671d77033b36be11fc20652c"
                    ],
                    [
                        {
                            "id": "ref"
                        },
                        "ref.fasta:md5,7315276c0e677b424e673562f48b9873"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,7cd4eb452407fcd946428f758c97442d"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-06-23T16:37:58.264534"
    },
    "HY - actinidia_chinensis - fasta - combinations": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "query_ref"
                        },
                        "query_ref.fasta:md5,7e52d0a17de86f5b11dd4d77c8332fa6"
                    ],
                    [
                        {
                            "id": "query"
                        },
                        "query.fasta:md5,bcc14495671d77033b36be11fc20652c"
                    ],
                    [
                        {
                            "id": "ref"
                        },
                        "ref.fasta:md5,7315276c0e677b424e673562f48b9873"
                    ]
                ],
                "1": [
                    "versions.yml:md5,0b0eba6b6f07300cd87c276e0a55424d",
                    "versions.yml:md5,5772594d0b3e1f41b2730f1dda23465b",
                    "versions.yml:md5,7cd4eb452407fcd946428f758c97442d",
                    "versions.yml:md5,c2aa26eec40ab4040601c089970c815c",
                    "versions.yml:md5,c8fc50b7f6a91b026265c5202a63fbc7"
                ],
                "fasta": [
                    [
                        {
                            "id": "query_ref"
                        },
                        "query_ref.fasta:md5,7e52d0a17de86f5b11dd4d77c8332fa6"
                    ],
                    [
                        {
                            "id": "query"
                        },
                        "query.fasta:md5,bcc14495671d77033b36be11fc20652c"
                    ],
                    [
                        {
                            "id": "ref"
                        },
                        "ref.fasta:md5,7315276c0e677b424e673562f48b9873"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0b0eba6b6f07300cd87c276e0a55424d",
                    "versions.yml:md5,5772594d0b3e1f41b2730f1dda23465b",
                    "versions.yml:md5,7cd4eb452407fcd946428f758c97442d",
                    "versions.yml:md5,c2aa26eec40ab4040601c089970c815c",
                    "versions.yml:md5,c8fc50b7f6a91b026265c5202a63fbc7"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.4"
        },
        "timestamp": "2025-07-02T10:18:26.935545"
    }
}